src/main/java/fr/openent/nextcloud/service/impl/DefaultDocumentsService.java

```diff
@@ -319,6 +319,68 @@ private JsonObject answerJsonError(Map.Entry<String, Future<JsonObject>> futureR
                 .put(Field.STATUS, Field.KO_LOWER);
     }
 
+    /**
+     * Create folder into workspace and copy every file from NC folder to new workspace folder.
+     * @param fileInfo      Information about the file
+     * @param file          Path to the file on nc server
+     * @param parentId      Identifier of the parent in the workspace (if you want to create the new folder under specific one)
+     * @param user          User infos
+     * @param userSession   Session infos
+     * @return              Infos about the copy
+     */
+    Future<JsonObject> folderCopy(JsonArray fileInfo, String file, String parentId, UserInfos user, UserNextcloud.TokenProvider userSession) {
+        Promise<JsonObject> promise = Promise.promise();
+        NextcloudFolder ncFolder = new NextcloudFolder(fileInfo);
+        if (!ncFolder.isSet()) {
+            String messageToFormat = "[Nextcloud@%s::folderCopy] Error while retrieve folder data : %s";
+            PromiseHelper.reject(log, messageToFormat, this.getClass().getSimpleName(), new Exception("folder.info.retrieve.failed"), promise);
+            return promise.future();
+        }
+        ncFolder.setPath(file);
+        EventBusHelper.createFolder(eventBus, ncFolder.getName(), parentId, user.getUserId(), user.getUsername())
+                .compose(folderInfos -> {
+                    ncFolder.setWorkspaceId(folderInfos.getString(Field.UNDERSCORE_ID));
+                    return copyDocumentToWorkspace(userSession, user, ncFolder.getFolderItemPath(), ncFolder.getWorkspaceId());
```

Comment:
In this function called by copyToWorkspace, we will call one more time copyDocumentToWorkspace. This is the recursive part of the function. The NextcloudFolder class takes care about the infinity loop possibilities.